tests/testthat/helper-functions.R

```diff
@@ -1,419 +1,422 @@
-  e <- try({
-    remDr$closeWindow()
-    remDr$closeServer()
-  }, silent = TRUE)
-  TRUE
-}
-
-# find the path to animint's testthat directory
-find_test_path <- function(dir = ".") {
-  dir <- normalizePath(dir, winslash = "/", mustWork = TRUE)
-  if (!grepl("animint", dir, fixed = TRUE))
-    stop("animint must appear somewhere in 'dir'")
-  base_dir <- basename(dir)
-  if (!base_dir %in% c("animint2", "tests", "testthat"))
-    stop("Basename of dir must be one of: 'animint2', 'tests', 'testhat'")
-  ext_dir <- switch(base_dir,
-                    animint2 = "tests/testthat",
-                    tests = "testthat",
-                    testthat = "")
-  file.path(dir, ext_dir)
-}
-
+translatePattern <-
+  paste0("translate[(]",
+         "(?<x>.*?)",
+         ",",
+         "(?<y>.*?)",
+         "[)]")
+
+acontext <- function(...){
+  print(...)
+  context(...)
+}
+
+## Parse the first occurance of pattern from each of several strings
+## using (named) capturing regular expressions, returning a matrix
+## (with column names).
+str_match_perl <- function(string,pattern){
+  stopifnot(is.character(string))
+  stopifnot(is.character(pattern))
+  stopifnot(length(pattern)==1)
+  parsed <- regexpr(pattern,string,perl=TRUE)
+  captured.text <- substr(string,parsed,parsed+attr(parsed,"match.length")-1)
+  captured.text[captured.text==""] <- NA
+  captured.groups <- do.call(rbind,lapply(seq_along(string),function(i){
+    st <- attr(parsed,"capture.start")[i,]
+    if(is.na(parsed[i]) || parsed[i]==-1)return(rep(NA,length(st)))
+    substring(string[i],st,st+attr(parsed,"capture.length")[i,]-1)
+  }))
+  result <- cbind(captured.text,captured.groups)
+  colnames(result) <- c("",attr(parsed,"capture.names"))
+  result
+}
+
+## Parse several occurances of pattern from each of several strings
+## using (named) capturing regular expressions, returning a list of
+## matrices (with column names).
+str_match_all_perl <- function(string,pattern){
+  stopifnot(is.character(string))
+  stopifnot(is.character(pattern))
+  stopifnot(length(pattern)==1)
+  parsed <- gregexpr(pattern,string,perl=TRUE)
+  lapply(seq_along(parsed),function(i){
+    r <- parsed[[i]]
+    starts <- attr(r,"capture.start")
+    if(r[1]==-1)return(matrix(nrow=0,ncol=1+ncol(starts)))
+    names <- attr(r,"capture.names")
+    lengths <- attr(r,"capture.length")
+    full <- substring(string[i],r,r+attr(r,"match.length")-1)
+    subs <- substring(string[i],starts,starts+lengths-1)
+    m <- matrix(c(full,subs),ncol=length(names)+1)
+    colnames(m) <- c("",names)
+    if("name" %in% names){
+      rownames(m) <- m[, "name"]
+    }
+    m
+  })
+}
+
+getSelectorWidgets <- function(html=getHTML()){
+  tr.list <- getNodeSet(html,
+                        '//table[@class="table_selector_widgets"]//tr')
+  td.list <- sapply(tr.list[-1], function(tr)xmlChildren(tr)[[1]])
+  sapply(td.list, xmlValue)
+}
+
+clickHTML <- function(...){
+  
+  v <- c(...)
+  
```

Comment:
please undo all of these additions of empty lines

Reply:
done